docs/reference-manual/api-reference.md

```diff
@@ -0,0 +1,43 @@
+# filter
+
+This section is testing znai integration with Doxygen.
+See below for some possible uses.
+
+> Seems like there is not much in terms of doc we can get for concepts?
+-----
+:include-doxygen-compound: fn::invocable_filter
+
+-----
+:include-doxygen-doc: fn::filter_t
+
+-----
+> Can't document niebloids (or even global variable) with this tool? The XML output of doxygen definitely contains this variable.
+<!-- :include-doxygen-compound: fn::filter -->
+<!-- :include-doxygen-doc: fn::filter -->
+
+-----
+:include-doxygen-compound: fn::filter_t
+
+:include-doxygen-compound: fn::filter_t::apply
+
+-----
+## All the operator() overloads
+
+> What i'd really like to be able to do is to grab all members of some entity and then use them as input to some template. So that we can generate similarly looking blocks of doc for each function or whatever.
+> I could not find a way to do it yet in Znai.
```

Comment:
This sounds like a good idea. Could you please elaborate?

Reply:
Certainly 👍 

I'd like something that allows me to create a "fragment" template, let's say to render documentation for one function. Then create another template for the class itself that iterates all member functions and renders them using the first "fragment" template.

In Freemarker it could look something like this.

- function.ftl:
```
- ${signature}
...
```

- class.ftl:
```
### ${name}
Member functions:
<#memberFunctions as mf>
    <@znai-import template="function.ftl" data=mf />
</#items>
---
Member variables:
<#memberVariables as mv>
    <@znai-import template="var.ftl" data=mv />
</#items>
```

- namespace.ftl:
```
## ${namespace.name}
<#namespace.classes as class>
    <@znai-import template="class.ftl" data=class />
</#items>
```

For this to work we would probably need to support something like `:include-doxygen-data: key` which simply returns **data** that then can be used with the templates. 

Hopefully the idea is more clear with the examples above.
Is this something that is already possible and i missed it? I have not been able to understand how to do it using the [templates in Znai's docs](https://testingisdocumenting.org/znai/layout/templates).

To recap, what i want is the ability to generate most of the documentation from Doxygen comments, not write the documentation by hand and only use Znai's Doxygen integration to get things like snippets of doc from individual functions and classes. It's very impressive how beautiful the documentation can be made with Znai but at this point it looks like i'd have to essentially write ALL of the documentation by hand, effectively copying the structure of the codebase instead of generating from it. Thanks for looking into this 🙇

Reply:
I like the idea to expand Znai templates to be more powerful. 
Template you are looking to build, how different they are going to be from Doxygen class compound (https://testingisdocumenting.org/znai/CPP/auto-reference#utilssecondanotherclass) ?

I do want to make default doxygen support to be viable so most people wouldn't need to generate their own template. 

Also I realize that Znai doesn't have everything you need right now and that can be annoying. Would love to use this opportunity to improve the doc system based on your feedback and features requests.

Reply:
The compound stuff is actually not bad at all. I have tried to use those and the output is pretty decent. 

The problem i have is that i can't automate the creation of a list of all classes in the namespace `fn` for example. Znai sort-of forces me to write the compound statement for each class by hand. 

I understand that this may be the way Znai was designed but i feel like you could add a few things here and there and it will allow us to generate things instead of explicitly list them. Just think about us adding a new class to `fn` namespace.. i'd like our CI to regenerate the beautiful doc through Znai instead of having to manually add the new class to the docs.. even if it's just a "compound" under a new heading.

Btw., nothing is annoying! 💙 I can just see how Znai can be improved. If you want to make it happen - great and Thank you. If not, then we will just have to write those by hand 👍 
Either way, thanks for taking the time to comment here!

Reply:
To provide more context, the idea behind manual specification of classes is so then you can move across files or group in a more meaningful way. Also separate some classes with diagrams or some other visuals. 

Wonder if adding `compound` for the namespace makes sense. But then it will generate a bunch of things in a single file...

Reply:
Yes, i see how that could be useful and i love the idea that there are essentially unlimited possibilities in terms of visuals (a very impressive set of features indeed!). 

I also realized the same problem - if we want to generate everything for a namespace it would either have to be all in one, potentially giant, file (not great and against Znai's ideas) or Znai would need to somehow support generating parts of the TOC using Freemarker or something like that (seems a bit overcomplicated). 

I think we can try to use Znai as it is and see how it goes. I'd also like to get it to work as part of our CI on this repo.
If you want a few easy improvements meanwhile:
- Automatically add `\n` at the end of generated files 
- Support non-capitalized labels/titles in TOC
    - In our library we have monadic operations such as `and_then`, `filter`, etc. and I'd like to have a top-level file for each of them. The title of such file would be `and_then`. But in the TOC it appears as `And_then` (automatic capitalization I assume). I'd like to be able to turn it off. Is there already a way that i missed?

Reply:
A few more observations while I did not forget:
- `noexcept` does not appear to be propagated in the signatures - this one is important i'd say
- same with attributes like `[[nodiscard]]` - this one might be ok, not a big deal, although i'd prefer to have those too
- adding a `@note` somewhere in the doxygen comment cuts the resulting documentation short (at least for `:include-doxygen-doc:`). i.e. the `@note` and anything behind it is not rendered

Reply:
These are good ones! Clearly our use case for C++ was way simpler. Adding GitHub issues and will prioritize:
https://github.com/testingisdocumenting/znai/issues/1180
https://github.com/testingisdocumenting/znai/issues/1181
https://github.com/testingisdocumenting/znai/issues/1182

Happy you tried the presentation mode! 
Do you have a suggestion about how to split into slides? I assume you want to use it to present and have a certain flow in mind

Reply:
> I'd also like to get it to work as part of our CI on this repo
what is missing? 
The crud way to get znai "binary" using wget from maven central: https://repo.maven.apache.org/maven2/org/testingisdocumenting/znai/znai-dist/1.67/znai-dist-1.67-znai.zip

there is a bit better integration with Maven build tool, but that is Java.

Close to ideal would be to have a GitHub action, but I haven't created any action so have a limited exposure. If this is something that you think could be useful - I will try to investigate

Reply:
Maven is unfortunately not of much use for us in C++.. 

> Close to ideal would be to have a GitHub action, but I haven't created any action so have a limited exposure. If this is something that you think could be useful - I will try to investigate

GitHub action would probably be the best. I'm happy to try embed Znai into our dev container and use it from CMake for now though. Have not tried yet. Will let you know if i run into any issues 👍 

> Do you have a suggestion about how to split into slides? I assume you want to use it to present and have a certain flow in mind

Did not really think this through yet - just thought it's a very cool feature and decided to try it out. I can see how it could be useful when pitching the library to others.

Perhaps a natural-ish flow would be to display (slide after slide)
- Name of the class and the text part (description of the class)
- All signatures on one slide if there are not many; or 
- Signature after signature if there are many (including parameters)

It's kinda hard to define what is "many" in this case as it also depends on the amount of text and parameters.. perhaps you can compute using total count of characters to keep it simple.

Reply:
This is helpful, thank you.  After my local fix `noexcept(true)` renders as `noexcept`, will this work? In generated doxygen XML I don't see a distinction between `noexcept(true)` and `noexcept`.

Re `[[nodiscard]]`, don't see anything related to it in my locally generated XML. Is there a chance you could take a look at your generated XML and check if there is any data related to it?

Reply:
In my case i do see some stuff like: `<highlight class="normal"><sp/><sp/>[[nodiscard]]<sp/></highlight>` in the XML output.

As for `noexcept(true)` vs `noexcept` - they are equivalent. 
What would be nice to output though is if in `noexcept(CONDITION)` the `CONDITION` is not a simple `true`. 
Idk. how well that is supported by Doxygen though.

Reply:
```<highlight class="normal"><sp/><sp/>[[nodiscard]]<sp/></highlight>``` is probably for the entire file content. There is another XML file where the args are parsed and there are also things like `const="yes"`. E.g:
```
      <memberdef kind="function" id="functional_8hpp_1ad744a75285a42e1302c54835d7298b4a" prot="public" static="no" const="no" explicit="no" inline="no" noexcept="yes" virt="non-virtual">
        <type>void</type>
        <definition>void repro_case::test1</definition>
        <argsstring>(int par1, std::string const &amp;par2) noexcept</argsstring>
        <name>test1</name>
        <qualifiedname>repro_case::test1</qualifiedname>
        <param>
          <type>int</type>
          <declname>par1</declname>
        </param>
```
That's what Znai parses

Reply:
Oh, in that case probably you just don't have the information to display attributes (e.g. `[[nodiscard]]`). To be honest this may be an oversight on Doxygen's side as I'm not really seeing these attributes in the HTML docs generated by Doxygen for another project either. As i said before - attributes would be _nice_ but are not crucial.